common/src/main/scala/net/psforever/objects/ShieldGeneratorDeployable.scala

```diff
@@ -1,15 +1,121 @@
+import net.psforever.objects.definition.{ComplexDeployableDefinition, SimpleDeployableDefinition}
 import net.psforever.objects.definition.converter.ShieldGeneratorConverter
+import net.psforever.objects.equipment.{JammableBehavior, JammableUnit}
+import net.psforever.objects.serverobject.PlanetSideServerObject
 import net.psforever.objects.serverobject.hackable.Hackable
+import net.psforever.objects.vital.Vitality
+import net.psforever.objects.zones.Zone
+import net.psforever.packet.game.PlanetSideGUID
+import services.avatar.{AvatarAction, AvatarServiceMessage}
+import services.Service
+import services.vehicle.{VehicleAction, VehicleServiceMessage}
 
 class ShieldGeneratorDeployable(cdef : ShieldGeneratorDefinition) extends ComplexDeployable(cdef)
   with Hackable
+  with JammableUnit
 
-class ShieldGeneratorDefinition extends DeployableDefinition(240) {
+class ShieldGeneratorDefinition extends ComplexDeployableDefinition(240) {
   Packet = new ShieldGeneratorConverter
   DeployCategory = DeployableCategory.ShieldGenerators
+
+  override def Initialize(obj : PlanetSideServerObject with Deployable, context : ActorContext) = {
+    obj.Actor = context.actorOf(Props(classOf[ShieldGeneratorControl], obj), s"${obj.Definition.Name}_${obj.GUID.guid}")
+  }
+
+  override def Uninitialize(obj : PlanetSideServerObject with Deployable, context : ActorContext) = {
+    SimpleDeployableDefinition.SimpleUninitialize(obj, context)
+  }
+}
+
+class ShieldGeneratorControl(gen : ShieldGeneratorDeployable) extends Actor
+  with JammableBehavior {
+
+  def JammableObject = gen
+
+  def receive : Receive = jammableBehavior
+    .orElse {
+      case Vitality.Damage(damage_func) =>  //note: damage status is reported as vehicle events, not local events
+        if(gen.Health > 0) {
+          val originalHealth = gen.Health
+          val cause = damage_func(gen)
+          val health = gen.Health
+          val damageToHealth = originalHealth - health
+          ShieldGeneratorControl.HandleDamageResolution(gen, cause, damageToHealth)
+          if(damageToHealth > 0) {
+            val name = gen.Actor.toString
+            val slashPoint = name.lastIndexOf("/")
+            org.log4s.getLogger("DamageResolution").info(s"${name.substring(slashPoint + 1, name.length - 1)}: BEFORE=$originalHealth, AFTER=$health, CHANGE=$damageToHealth")
+          }
+        }
+
+      case _ => ;
+    }
+
+  override def StartJammeredSound(target : Any, dur : Int) : Unit =  { }
+
+  override def StartJammeredStatus(target : Any, dur : Int) : Unit = target match {
+    case obj : PlanetSideServerObject with JammableUnit if !obj.Jammed =>
+      obj.Zone.VehicleEvents ! VehicleServiceMessage(obj.Zone.Id, VehicleAction.PlanetsideAttribute(Service.defaultPlayerGUID, obj.GUID, 27, 1))
+      super.StartJammeredStatus(obj, dur)
+    case _ => ;
+  }
+
+  override def CancelJammeredSound(target : Any) : Unit =  { }
```

Comment:
I assume this and `StartJammeredSound` are empty because it has no jammered sound, or is this an oversight?

If the former, perhaps a comment might be useful for anyone looking at it in the future.

Reply:
As mentioned in Caveat 2, the shield generator is called out in the ADB entries for jammer projectiles but it has no obvious responses to the conventional jammering packets being received.  It is also a single unit at this moment of implementation.

For example, when the vehicle is jammered, its weapons are also jammered.  Both elements emit the buzz sound but jammering the weapons is required to make them stop shooting.  The Aegis Shield Generator has no need to have any subcomponents being jammed because it has no subcomponents at the moment, hence, no reason to be jammered twice.  Overriding the `Start...` method stops the timer from being started; overriding the `Cancel...` is just for completion.  It didn't matter whether it was `Sound` or `Status`.

Reply:
It might be worth adding a small explanation of that as a comment in that case, so for future reference we remember why it's overriden.

Reply:
Commented.